RFCs/rfc-0004-fast-transactions.md

```diff
@@ -0,0 +1,77 @@
+# RFC-0004: Decreasing the minimum interval between SmartContract's Transactions from 3 minutes to 1 second with Recursive Proofs
+
+- **Intent**: Implement recursive proofs in SmartContract transactions to reduce the minimum transaction interval from 3 minutes to 1 second.
+- **Submitted by**: DFST ([Github](https://github.com/dfstio), [Twitter](https://twitter.com/dfst_io), Discord: dfst.io)
+- **Date**: Thursday, December 07, 2023
+
+## Abstract
+
+This RFC aims to improve the Mina Protocol's developer and user experience by enabling faster transaction processing in SmartContracts. Currently, a minimum interval of 3 minutes is required between SmartContract transactions that have preconditions and alter the state. This proposal intends to shorten this interval to 1 second, allowing rapid transaction execution with state preconditioning and changing.
```

Comment:
This is related to not just smart contracts but more generally block-time, correct?

Reply:
You are correct, but other use cases are already possible without this RFC. You can now calculate recursive proofs with ZkProgram and include many proofs in one block. You can send many non-contract transactions by increasing the nonce by one each time, and all those transactions will be included in the same block.

What is different with SmartContract is that to include the transaction to the block, the precondition is checked. This precondition now is checked against the last block account state, so you cannot include several transactions that rely on different precondition states, you should wait for the next block.

Reply:
The zkVM that @teddyjfpender described above is already implemented by Protokit, you can start using it today at https://protokit.dev

Realistically we can do between 3-12 TPS with some minor changes to the underlying protocol, we can also do *instant* transaction inclusion confirmation (via receipts), data availability via sequence state or events and many other things that can be added to the core protocol.

Features available currently are:
- Abstractions over writing zk programs that manage state via state transitions, 
- Soft-failing transactions, inclusion of failed transaction in a provable way into a block
- Key-value storage a.k.a. "infinite" state
- Out-of-prover sequencing of transactions
- Parallelisation of proof generation in block production (by separating tree operations and business logic)
- Development environment with support for UI - via GQL endpoints and a handy CLI

Features that are currently being implemented:
- L1/L2 bridge
- Sequence state mempool, as opposed to the private mempool used currently
- Indexing & processing of historical data
- and many more features coming soon! :)

I'd encourage anyone facing the issues described in the RFC to try out our starter kit and join our discord, here's a list of useful links and resources:
- https://github.com/proto-kit/starter-kit
- https://protokit.dev
- https://discord.com/invite/AMGnGAxsKp
- https://www.youtube.com/watch?v=hqobPb3gIRo

Reply:
I've included the chart above in the new version of RFC. Thank you, Teddy, for it!
I've also added several new diagrams to illustrate RFC concepts.

The information about possible TPS above is very interesting, I've included the section "Lower theoretical bound of the interval between transactions and why it matters" to the RFC to discuss the lower possible bound of the interval.